### PR TITLE
Feature/core 631 identity based signer

### PR DESCRIPTION
**Pull request by wulfraem**
- related to https://github.com/evannetwork/smart-contracts-core/pull/2
- [CORE-631]

[CORE-631]: https://evannetwork.atlassian.net/browse/CORE-631

### CODE REVIEW

File: src/common/utils.ts [hunk 1]
```diff
@@ -20,6 +20,9 @@
 import { Runtime } from '../index'
 
 
+export const nullAddress = '0x0000000000000000000000000000000000000000';
```
Comment (karladler) [inline src/common/utils.ts]: shall we use caps-lock style for definition constants? NULL_ADDRESS and NULL_BYTES_32 ?

File: src/contracts/signer-identity.spec.ts [hunk 2]
```diff
@@ -0,0 +1,364 @@
+/*
+  Copyright (c) 2018-present evan GmbH.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (karladler) [inline src/contracts/signer-identity.spec.ts]: when do we actually use the apache license?

File: src/contracts/signer-identity.spec.ts [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+      {
+        activeIdentity: await verifications.getIdentityForAccount(accounts[3], true),
+        underlyingAccount: accounts[3],
+        underlyingSigner,
+      },
+    );
+    executor = new Executor(
+      { config: { alwaysAutoGasLimit: 1.1 }, signer: signer, web3 });
+    await executor.init({ eventHub: await TestUtils.getEventHub(web3) });
+
```
Comment (karladler) [inline src/contracts/signer-identity.spec.ts]:
```suggestion
```

File: src/contracts/signer-identity.spec.ts [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+          await expect( executor.executeSend({
+            from: signer.activeIdentity,
+            to: contract.options.address,
+            gas: 100e3,
+            value: amountToSend,
+          })).to.rejected;
+
+          const balanceAfter = new BigNumber(await web3.eth.getBalance(signer.underlyingAccount));
+          expect(balanceAfter.eq(balanceBefore)).to.be.not.true;
```
Comment (karladler) [inline src/contracts/signer-identity.spec.ts]: not sure about this code, but looks better to me than testing for .eq to be not true ;) 

```suggestion
          expect(balanceAfter).to.be.lt(balanceBefore);
```

Comment (karladler) [reply src/contracts/signer-identity.spec.ts]: check also this style for the other tests

Comment (karladler) [reply src/contracts/signer-identity.spec.ts]: ah I see, you implemented it like this, bc of BigNumber?

File: src/contracts/signer-identity.ts [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+   */
+  public async createContract(
+    contractName: string,
+    functionArguments: any[],
+    options: any,
+  ): Promise<any> {
+    if (options.from === this.underlyingAccount) {
+      return this.config.underlyingSigner.createContract(contractName, functionArguments, options);
+    }
+    // build input for contructror call
```
Comment (karladler) [inline src/contracts/signer-identity.ts]:
```suggestion
    // build input for constructor call
```

File: src/contracts/signer-identity.ts [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+      this.encodeConstructorParams(JSON.parse(compiledContract.interface), functionArguments);
+
+    const { blockNumber, transactionHash } = await this.handleIdentityTransaction(null, null, [], options);
+    const keyHolderLibrary = this.options.contractLoader.loadContract(
+      'KeyHolderLibrary', this.activeIdentity);
+    const events = await keyHolderLibrary.getPastEvents(
+      'ContractCreated', { fromBlock: blockNumber, toBlock: blockNumber });
+    const matches = events.filter(ev => ev.transactionHash === transactionHash);
+    if (matches.length !== 1) {
+      throw new Error('contract creation failed');
```
Comment (karladler) [inline src/contracts/signer-identity.ts]: in which situations could this be the case?Can we make the error message more specific?

File: src/contracts/signer-identity.spec.ts [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+          const randomString = Math.floor(Math.random() * 1e12).toString(36);
+          const contract = await executor.createContract(
+            'TestContract', [randomString], { from: signer.activeIdentity, gas: 1e6 });
+          expect(contract).to.be.a('Object');
+
+          const amountToSend = Math.floor(Math.random() * 1e3);
+
+          const balanceBefore = new BigNumber(await web3.eth.getBalance(signer.underlyingAccount));
+
+          await expect( executor.executeSend({
```
Comment (S3bb1) [inline src/contracts/signer-identity.spec.ts]:
```suggestion
          await expect(executor.executeSend({
```

File: src/contracts/signer-identity.spec.ts [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+            from: signer.activeIdentity,
+            to: contract.options.address,
+            gas: 100e3,
+            value: amountToSend,
+          })).to.rejected;
```
Comment (S3bb1) [inline src/contracts/signer-identity.spec.ts]:
```suggestion
          })).to.be.rejected;
```

File: src/contracts/signer-identity.spec.ts [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+        const contract = await executor.createContract(
+          'TestContract', [], { from: signer.activeIdentity, gas: 1e6 });
+        expect(contract).to.be.a('Object');
+        const balanceBefore = new BigNumber(await web3.eth.getBalance(contract.options.address));
+        await executor.executeContractTransaction(
+          contract, 'chargeFunds', { from: signer.activeIdentity, value: amountToSend });
+        const balanceAfter = new BigNumber(await web3.eth.getBalance(contract.options.address));
+        const diff = balanceAfter.minus(balanceBefore);
+        expect(diff.eq(new BigNumber(amountToSend))).to.be.true;
+        expect(balanceAfter.eq(balanceBefore)).to.be.not.true;
```
Comment (S3bb1) [inline src/contracts/signer-identity.spec.ts]:
```suggestion
        expect(balanceAfter.eq(balanceBefore)).to.be.false;
```

File: src/contracts/signer-identity.spec.ts [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+        expect(contract).to.be.a('Object');
+
+        const amountToSend = Math.floor(Math.random() * 1e3);
+        const balanceBefore = new BigNumber(await web3.eth.getBalance(contract.options.address));
+        await expect( executor.executeSend({
+          from: signer.activeIdentity,
+          to: contract.options.address,
+          gas: 100e3,
+          value: amountToSend,
+        })).to.rejected;
```
Comment (S3bb1) [inline src/contracts/signer-identity.spec.ts]:
```suggestion
        })).to.be.rejected;
```

File: src/contracts/signer-identity.spec.ts [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+          expect(contract).to.be.a('Object');
+
+          const amountToSend = Math.floor(Math.random() * 1e3);
+          const balanceBefore = new BigNumber(await web3.eth.getBalance(signer.activeIdentity));
+          await expect( executor.executeSend({
+            from: signer.activeIdentity,
+            to: contract.options.address,
+            gas: 100e3,
+            value: amountToSend,
+          })).to.rejected;
```
Comment (S3bb1) [inline src/contracts/signer-identity.spec.ts]:
```suggestion
          })).to.be.rejected;
```

File: src/contracts/signer-identity.ts [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+  public async signAndExecuteSend(
+    options: any,
+    handleTxResult: Function,
+  ): Promise<void> {
+    if (options.from === this.underlyingAccount) {
+      return this.config.underlyingSigner.signAndExecuteSend(options, handleTxResult);
+    }
+
+    try {
+      debugger;
```
Comment (S3bb1) [inline src/contracts/signer-identity.ts]:
```suggestion
```

File: src/common/utils.ts (hunk 1 above)
Comment (wulfraem) [reply src/common/utils.ts]: Usual code style does not use upper case constants. We had a JavaScript code style, that forbade this. We carried over this convention when moving to TypeScript.

File: src/contracts/signer-identity.spec.ts (hunk 3 above)
Comment (wulfraem) [reply src/contracts/signer-identity.spec.ts]: Empty lines at the end of files is an older but broadly used POSIX/Unix convention. We have this in most api project files.

File: src/contracts/signer-identity.ts (hunk 6 above)
Comment (wulfraem) [reply src/contracts/signer-identity.ts]: Unfortunately not, contract creation may fail due to some inner transactions, that failed. A more detailed reason can only be found by manual analysis.

File: src/contracts/signer-identity.spec.ts (hunk 4 above)
Comment (wulfraem) [reply src/contracts/signer-identity.spec.ts]: Updated tests to use `.lt` and `.gt`, currently there's no real need for chai-bignumber, I think ;)

File: src/contracts/signer-identity.spec.ts (hunk 2 above)
Comment (wulfraem) [reply src/contracts/signer-identity.spec.ts]: Fixed license headers.

File: src/contracts/signer-identity.spec.ts (hunk 9 above)
Comment (wulfraem) [reply src/contracts/signer-identity.spec.ts]: Line has been replaced by other updates.

**Pull request merged**